### ISSUE TITLE
Return value for `pact_verifier.verify`

### ISSUE THREAD

**Issue by Mukheem** [reecetech/pactman#85]
Hey...!!! 

I was doing a provider test using pytest and shared local repository instead of pact broker. In this case I found that my test is passed irrespective of the actual result.

Because the following code returns nothing. `pact_verifier.verify(live_server.url, provider_state)` 

How could this be asserted properly without a slip during actual tests ??

Related code holds the result(success/failure) but returns nothing. Perhaps, It is designed to cater PACT broker. For all the audience who are interested to implement pactman without using pact broker,How is this going to help !!!
Can you change the code to provide a return type ??

If you feel changing this code hampers a lot of other features, then you may provide another method which would do the same thing but return the response along with status. (Return type: dict)

**Comment by ryallsa** [comment]
Thanks for contacting Registers of Scotland, your email has been delivered to your intended recipient(s), who will be in touch as soon as they can.

In order to focus on safeguarding the wellbeing of our staff, maintaining our essential business activity, and operating in accordance with government guidelines during this public health emergency, we ask that you direct your enquiries as follows:

A specific case or enquiry – please send all enquiries relating to a specific case to our  Customer Services<mailto:customer.services@ros.gov.uk> team, who will be in touch.

COVID 19 general enquiry – our website is regularly updated with all the latest information and updates<https://www.ros.gov.uk/coronavirus-covid-19> in relation to Coronavirus. If you have a COVID-19 related enquiry that is not answered on our webpage then please email Covid19enquiries@ros.gov.uk<mailto:Covid19enquiries@ros.gov.uk>, where your enquiry will be directed to the most appropriate/available person to respond.

Any other general enquiry – we would encourage you to continue emailing mailboxes associated with our departments or business functions, rather than emailing specific individuals. If you have concerns that your intended recipient has not received your enquiry then please resend your email to Covid19enquiries@ros.gov.uk<mailto:Covid19enquiries@ros.gov.uk> so it can be directed to the most appropriate/available person to respond.

Thank You

- - - - - - - - - - - - - - - - - - -
Registers of Scotland
- - - - - - - - - - - - - - - - - - -

www.ros.gov.uk<https://www.ros.gov.uk>

[cid:imageea4625.PNG@0ff942e5.4392717d]<https://www.ros.gov.uk/>    [cid:image4c7582.PNG@c6961eb6.4b8d12dc] <https://www.facebook.com/registersofscotland/>   [cid:image3d44cf.PNG@3d3dc956.4a9cbb28] <https://twitter.com/RegistersOfScot>   [cid:image7a2388.PNG@18aac01a.4c8df857] <https://www.youtube.com/channel/UCsvdb1Mw7n-shANwtFxrDMQ>   [cid:image512ca4.PNG@e374260f.429e7de7] <https://www.linkedin.com/authwall?trk=gf&trkInfo=AQEFPBtu7CCWwQAAAWNewfqoDlR1tdKRE7wGUu67suRzsyti8TMecqKnr1l97Cm06GPZkKXX2TXzf5g1bQO13VHKbYipr9Tf136ULQZRIGWkYOfOjp8PupaH9JDvdmTdxeUFykI=&originalReferer=&sessionRedirect=https%3A%2F%2Fwww.linkedin.com%2Fcompany%2Fregisters-of-scotland>

Please consider the environment – do you really need to print this email?

This e-mail (and any files or other attachments transmitted with it) is intended solely for the attention of the addressee(s). Unauthorised use, disclosure, storage, copying or distribution of any part of this e-mail is not permitted. If you are not the intended recipient please destroy the email, remove any copies from your system and inform the sender immediately by return. 

Registers of Scotland does not accept any liability or responsibility for any damage caused by any virus transmitted by this email or for changes made to this e-mail after it was sent. 
All communications to or from Registers of Scotland may be automatically logged, monitored and/or recorded in order to secure the effective operation of the system and for other lawful purposes. The views or opinions contained within this e-mail may not necessarily reflect those of the Keeper of the Registers of Scotland.

**Comment by r1chardj0n3s** [comment]
If the pact verification fails then an assertion should be raised by the verify() code. Are you seeing a verification failure? This isn't really anything to do with the pact broker.

**Comment by Mukheem** [comment]
Thanks @r1chardj0n3s  for replying.

If the pact verification fails then an assertion should be raised by the verify() code. --> **Yes, This is what we need.**We need some sort of assertion/return value to be thrown by` .verify `method

Are you seeing a verification failure?  --> 
I'm seeing a verification failure provided by pactman but pytest test case passes because the failure isn't actually caught by pytest. ( because failure is simply a log from pactman which is bring printed.)

This needs a human intervention to identify if the test is actually passed or failed. 

If this has nothing to do with pact broker then how are we gonna publish the results using a local shared directory ?

**Comment by r1chardj0n3s** [comment]
Just to be clear: what's the log you're seeing that indicates a failure, but doesn't result in the assertion being raised?

**Comment by Mukheem** [comment]
@r1chardj0n3s 

First line in the image shows that pytest test case is passed. where as the actual issue is shown in the second line. i.e., pactman log showing that provider test failed.


![image](https://user-images.githubusercontent.com/19348206/80343383-6718e700-8883-11ea-808f-9bdf4d2d5996.png)


Second line is followed by a detailed log what is expected and what is returned. 

The problem here is Unless some one looks into the log, pytest makes us believe that the test is passed. :(

**Comment by r1chardj0n3s** [comment]
Hmm. That should be resulting in a failure being thrown by verify(). I'd need to know what sort of failure it is to be able to debug why it's not being thrown though.

**Comment by Mukheem** [comment]
The case which I was testing is 'When one of the values does not match the response data provided'

But ideally, what ever the case might be, it should throw up failure such that test case catches it, @r1chardj0n3s .

**Comment by r1chardj0n3s** [comment]
I KNOW it should be doing that! I don't know how many times I need to say that!

There is a bug inside verify which is causing it NOT to raise the test failure. I'm trying to determine what the actual failed pact assertion was that is not raising the test failure. I need more information from you regarding exactly what pact assertion you're seeing. Could you please run with verbose reporting and give me that detail?

**Comment by r1chardj0n3s** [comment]
To be clearer: the "verbose reporting" is turned on with pytest's `-v` option.

**Comment by Mukheem** [comment]
Hey @r1chardj0n3s ... Here is the detailed report.

```
PASSED                                                                   [100%]
Consumer: *****
Request: "Fetches details of the given ***** id" ... FAILED
 Setting up provider state 'There should be at least one existing *****'
 Verifying v2 Response: status=200, headers=absent, body=present, matching rules=present
 check_rules data={'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'expiresAt': None, 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'respondedAt': None, 'notInterestedAt': None, 'notInterestedReason': None, 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'TEXAS', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'} spec={'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'Texas', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'} path=body
 apply_rules data={'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'expiresAt': None, 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'respondedAt': None, 'notInterestedAt': None, 'notInterestedReason': None, 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'Texas', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'} spec={'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'Texas', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'} path=body
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 apply_rules_dict data={'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'expiresAt': None, 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'respondedAt': None, 'notInterestedAt': None, 'notInterestedReason': None, 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'Texas', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'} spec={'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'Texas', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'} path=body
 apply_rules data='interested' spec='interested' path=body.status
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'status'] got weights [(8, '$.body.status')]
 ... rule lookup got rule="Rule match by {'match': 'type'} at $.body.status", weight=8
 match type 'interested' 'interested' ['body', 'status']
 ... passed
 apply_rules data='2019-05-18T06:57:35+00:00' spec='2019-05-18T06:57:35+00:00' path=body.createdAt
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'createdAt'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='2019-05-28T07:45:11+00:00' spec='2019-05-28T07:45:11+00:00' path=body.initialView
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'initialView'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='2019-05-28T07:58:09+00:00' spec='2019-05-28T07:58:09+00:00' path=body.showDetailsAt
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'showDetailsAt'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='direct' spec='direct' path=body.requestType
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'requestType'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='C.F.S. Pest Control' spec='C.F.S. Pest Control' path=body.*****
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', '*****'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='Texas' spec='Texas' path=body.*****
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', '*****'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='3000' spec='3000' path=body.*****
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', '*****'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='--------------------------------' spec='--------------------------------' path=body.chatUrl
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'chatUrl'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='--------------------------------' spec='--------------------------------' path=body.consumerDetailName
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'consumerDetailName'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules data='TX' spec='TX' path=body.consumerDetailLocation
 find_rule got [<MatchType path='$.body.status' rule={'match': 'type'}>] for section body
 ... path ['$', 'body', 'consumerDetailLocation'] got weights [(0, '$.body.status')]
 ... rule lookup got None
 ... falling back on equality matching
 apply_rules body success=False
 check_rules success=False
 Response data was: {'status': 'interested', 'createdAt': '2019-05-18T06:57:35+00:00', 'expiresAt': None, 'initialView': '2019-05-28T07:45:11+00:00', 'showDetailsAt': '2019-05-28T07:58:09+00:00', 'respondedAt': None, 'notInterestedAt': None, 'notInterestedReason': None, 'requestType': 'direct', '*****': 'C.F.S. Pest Control', '*****': 'Texas', '*****': '3000', 'chatUrl': '--------------------------------', 'consumerDetailName': '--------------------------------', 'consumerDetailLocation': 'TX', 'consumerDetailEmail': '--------------------------------', 'consumerDetailPhone': '--------------------------------', 'detailsBudget': '10000', 'detailsTimeframe': 'Urgent', 'detailsNeeds': 'Cockroaches everywhere!'}```

Apologies if my words irked you. :)

**Comment by r1chardj0n3s** [comment]
Thanks for that! I won't be able to look at it in detail for a few hours, so if you're able to correct the formatting between now and then I'd appreciate it ;-)

**Comment by Mukheem** [comment]
I would love to before you comment. but unfortunately that's what i also see. what I could suggest is to copy the content and paste in notepad or notepad++ unwrap the files. 
That would give you a better insight to read thru it.

That's what i also did. :) @r1chardj0n3s

**Comment by r1chardj0n3s** [comment]
FYI the markdown triple-back-quote is required to mark a large block of preformatted text. I've gone ahead and edited your comment to do that.

The verbose output really isn't very user-friendly, I wish I could do better there :(

I've looked into it though and I can't see any reason why it's saying `apply_rules body success=False` in that output. All the actual verification checks appear to pass. I'm perplexed! :)

Could I get you to send me the actual pact test code? That way I can reproduce the problem and get right into why it's happening. Please sanitise it as much as you need to as long as it still fails. Feel free to send it to my gmail (my github username and gmail address are the same) if you are worried about information being public.

**Comment by Mukheem** [comment]
Thanks @r1chardj0n3s  for doing it on my behalf. I never knew that before. 

I shall update you in some time. I need a while to sanitize the code.

**Comment by Mukheem** [comment]
Hey @r1chardj0n3s ,

I sent you an email with all the related details.

**Comment by r1chardj0n3s** [comment]
Yep, got it. It's Friday evening here, so I'll take a look tomorrow.

**Comment by r1chardj0n3s** [comment]
Based on the code you've provided, you re-wrote part of the verification code to be able to test a single pact file against a remote provider using pytest. FWIW there's no point - you should just use the `pactman-verifier` command line with the `-l` option since you're getting no benefit from running inside pytest.

The real issue you're running into though is that you're using `CaptureResult` instead of `PytestResult` so the pytest `Failed` exception isn't being raised on a failure.

**Comment by Mukheem** [comment]
We are using single file just because we are still in development phase. But going further we will have n number of pacts and command line would not be the best choice.

Can you help me with the usage of `PytestResult`. Because it is quite necessary that we catch the failure in pytest test case.

**Comment by r1chardj0n3s** [comment]
In your code where you have CaptureResult, replace it with PytestResult. It will cause the pytest test to fail when the verification fails.

**Comment by r1chardj0n3s** [comment]
ps. you really should look into setting up a broker, it's a really good idea.

**Comment by Mukheem** [comment]
@r1chardj0n3s , I see `PytestResult` is passing the test even when one of the JSON values is different by one character. Where as the same test with `CaptureResult` is failing.

Usage of broker leads us against one of the organization's policies. Hence We cannot use that.

**Comment by r1chardj0n3s** [comment]
The failure I saw was that the type of the "id" JSON field was wrong: you expected a string, but received a number. The Result class used does not affect any of the tests. The "different by one character" you mention is not the failure I saw in the sample you gave me. Also, I recall the data being compared by type, so a one character difference won't cause a failure - the values are of the same type.

Are you aware you can run the broker within your organisation? You don't have to use the hosted one.

**Comment by Mukheem** [comment]
@r1chardj0n3s ,

When we have no matchers used, even one character failure should should be reported.

Ain't I correct ?

Also, the same test with `PytestResut` is passing when `CaptureResult` is failing. Why is there such difference.. I just need this to make my tests robust and leak proof.

**Comment by r1chardj0n3s** [comment]
When you use no matchers then pact operates on an exact value match basis, so a difference in content should result in a failure.

Please send me the new code.

**Comment by Mukheem** [comment]
Sent it, @r1chardj0n3s 

:)

**Comment by r1chardj0n3s** [comment]
Thanks for your patience. I have tracked down the bug, and the fix is released as 2.28.0

FWIW I would still recommend that if you're just using pact files, then just use the pactman-verifier command line tool. If you're going to be using pact seriously, you really should use a broker that the consumer and provider sides can use to manage the pacts between them, including seeing historical results and being able to easily query whether pacts are currently passing or not.

**Comment by Mukheem** [comment]
Thanks @r1chardj0n3s  for fixing it. We see it working fine. Closing the issue. :)
